[engyuhan] iP

engyuhan opened the pull request:
# Bob

> “Bob is cool.” – Barack Obama ([source](https://barackobama.com/))

Bob is cool. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/engyuhan/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon) 
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

engyuhan committed:
refactor: rename to Bob

engyuhan committed:
feat: add greet and exit

engyuhan committed:
feat: add echo

engyuhan committed:
feat: add list

engyuhan committed:
refactor: add ChatBot class

engyuhan committed:
docs: add documentation for ChatBot class

engyuhan committed:
feat: add InvalidInputException class

engyuhan committed:
feat: add Task interface

engyuhan committed:
feat: add BasicTask class

engyuhan committed:
feat: change ChatBot list to tasks, add mark command

engyuhan committed:
feat: change Task to concrete class to replace BasicTask

engyuhan committed:
feat: add ToDo class

engyuhan committed:
feat: add Deadline class

engyuhan committed:
feat: add Event class

engyuhan committed:
feat: add todo, deadline and event methods to ChatBot to replace add method

engyuhan committed:
feat: add todo, deadline and event commands

engyuhan committed:
test: add text-ui-test

engyuhan committed:
fix: fix bug where invalid input for mark and unmark commands crashes program

engyuhan committed:
feat: add delete method to ChatBot

engyuhan committed:
feat: add delete command

engyuhan committed:
test: add test for delete command

engyuhan committed:
refactor: Add Tasks class

engyuhan committed:
refactor: Add tasks package

engyuhan committed:
fix: Fix bug with tasks where InvalidInputException is not imported.

engyuhan committed:
feat: Add SavedTasks class

engyuhan committed:
feat: Add save functionality to Tasks class

engyuhan committed:
feat: Add save

engyuhan committed:
feat: Change timings to LocalDateTime in task classes

engyuhan committed:
feat: Change SavedTasks to parse timings into LocalDateTime

engyuhan committed:
feat: Change ChatBot to parse input timings into LocalDateTime

engyuhan committed:
feat: Add Command class

engyuhan committed:
feat: Add AddCommand class

engyuhan committed:
feat: Add DeleteCommand class

engyuhan committed:
feat: Add ListCommand class

engyuhan committed:
feat: Add MarkCommand class

engyuhan committed:
feat: Add ExitCommand class

engyuhan committed:
feat: Add NoopCommand class

engyuhan committed:
refactor: Move task package into duke package

engyuhan committed:
refactor: Rename ChatBot to Ui, update functionality to only reading and writing messages

engyuhan committed:
refactor: Rename SavedTasks to Storage, update functionality to only interacting with persistent storage

engyuhan committed:
refactor: Rename Tasks to TaskList, update functionality to only interacting with list of tasks

engyuhan committed:
refactor: Add Parser class

engyuhan committed:
refactor: Rename InvalidInputException to BobException

engyuhan committed:
refactor: Update Bob class for refactored classes

engyuhan committed:
Merge branch 'add-gradle-support'
feat: Add gradle support

engyuhan committed:
build: Add checkstyle

engyuhan committed:
style: Updated style to comply to style check

engyuhan committed:
test: Add tests for tasks

engyuhan committed:
build: Add shadowJar to gradle

engyuhan committed:
style: Update tests to comply with style guide

engyuhan committed:
docs: Add documentation for Bob class

engyuhan committed:
feat: Add FindCommand class

engyuhan committed:
feat: Add find method to Ui

engyuhan committed:
feat: Add find message to Parser

engyuhan committed:
Merge branch 'branch-A-CodingStandard'
style: Update tests to comply with code style

engyuhan committed:
Merge branch 'branch-level-9'
feat: Add find command

engyuhan committed:
build: Add check style

engyuhan committed:
feat: Update Ui to use input and output message queues

engyuhan committed:
feat: Update Bob to use input and output message queues, and provide gui option

engyuhan committed:
feat: Add formatter for ui

engyuhan committed:
feat: Add gui

daronoh commented on src/main/java/duke/Bob.java:
```diff
@@ -0,0 +1,94 @@
+    public static void main(String[] args) {
+        BlockingQueue<String> inputQueue = new ArrayBlockingQueue<>(10);
+        BlockingQueue<String> outputQueue = new ArrayBlockingQueue<>(10);
+        if (isGui) {
```
Would it be better to space out and group lines of code based on their purpose?

daronoh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,246 @@
+        String response;
+        String command = input.split(" ")[0];
+        return switch (command) {
+        case "bye" -> new ExitCommand();
+        case "list" -> new ListCommand();
+        case "mark" -> mark(input);
+        case "unmark" -> unmark(input);
+        case "todo" -> todo(input);
+        case "deadline" -> deadline(input);
+        case "event" -> event(input);
+        case "delete" -> delete(input);
+        case "find" -> find(input);
+        default -> throw new BobException("I'm sorry, I did not understand your message.");
```
should these be done as one-liners or egyptian style?

daronoh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,246 @@
+            throw new BobException("Please specify which task to mark.");
+        }
+
+        int itemNum = Integer.parseInt(matcher.group(1));
```
is itemNum abit too ambiguous? might be better to name it something like itemIndex.

daronoh commented on src/main/java/duke/command/NoopCommand.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Represents a no operation command.
+ */
+public class NoopCommand extends Command {
```
perhaps a better name for NoopCommand might just be NoCommand?

Xczheng0105 replied:
Could consider InvalidCommand instead of NoopCommand as a more intuitive name here.

daronoh reviewed (commented):
LGTM. Nice Work!

Xczheng0105 commented on src/main/java/duke/Bob.java:
```diff
@@ -0,0 +1,94 @@
+public class Bob {
+    private static final boolean isGui = true;
+    private TaskList tasks;
+
```
Do remove the empty lines.

Xczheng0105 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,246 @@
+     * @throws BobException If the input is invalid.
+     */
+    private static Command unmark(String input) throws BobException {
+        Matcher matcher = Pattern.compile("^unmark (\\d*)$").matcher(input);
```
Do consider a more descriptive name for "Matcher" such as CommandMatcher. Alternatively, you could use something like "Parser".

Xczheng0105 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,130 @@
+            if (lineNum == itemNum) {
+                Task task = Parser.parseStorage(line);
+                task.mark(isCompleted);
+                writer.write(task + System.getProperty("line.separator"));
```
I like the use of line separator here =)

Xczheng0105 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @param msg The message to format.
+     */
+    public void write(String msg) {
```
Once again, you could consider more descriptive names like readMessage or writeMessage here.

Xczheng0105 reviewed (commented):
No major issues generally. LGTM. Would be good if some of the names were made more intuitive.

engyuhan committed:
refactor: Extract parse methods and mark and delete methods in storage

Parser::parseStorage, Storage::mark and Storage::delete are long methods

Extracting code fragments into separate methods make the code more readable.

engyuhan committed:
feat: Add assertions

All methods use no assertions.

Assertions can be used for check assumptions are upheld in code.

engyuhan committed:
feat: Use streams

engyuhan committed:
Merge pull request #1 from engyuhan/branch-A-CodeQuality

Code Quality

engyuhan committed:
Merge branch 'master' into branch-A-Assertions

engyuhan committed:
Merge pull request #2 from engyuhan/branch-A-Assertions

Assertions

engyuhan committed:
Merge branch 'master' into branch-A-Streams

engyuhan committed:
Merge pull request #3 from engyuhan/branch-A-Streams

Streams

engyuhan committed:
Merge branch 'master' of github.com-engyuhan:engyuhan/ip into branch-A-Streams

engyuhan committed:
build: Add github workflows

engyuhan committed:
Merge pull request #4 from engyuhan/branch-A-CI

Add Github Workflows

engyuhan committed:
feat: Create Archive

engyuhan committed:
feat: Add add multiple and clear methods to Storage

engyuhan committed:
feat: Add add multiple and clear methods to TaskList

engyuhan committed:
feat: Add archive and restore methods to Ui

engyuhan committed:
feat: Create ArchiveCommand

engyuhan committed:
feat: Create RestoreCommand

engyuhan committed:
feat: Update commands to take an archive parameter

engyuhan committed:
feat: Add archive and restore commands to Parser

engyuhan committed:
feat: Update Bob to use archive

engyuhan committed:
Merge pull request #5 from engyuhan/BCD-Extension

Archive

engyuhan committed:
feat: Add user guide

engyuhan committed:
fix: Fix file storage
